tests/reprsimil/test_brsa.py

```diff
@@ -231,113 +264,219 @@ def test_gradient():
     snr = np.exp(np.dot(L,np.random.randn(n_V))) * snr_level
     sqrt_v = noise_level*snr
     betas_simulated = np.dot(L_full,np.random.randn(n_C,n_V)) * sqrt_v
-    signal = np.dot(design.design_used,betas_simulated)
+    signal = np.dot(design.design_task,betas_simulated)
 
     # Adding noise to signal as data
     Y = signal + noise
 
 
-    scan_onsets = np.linspace(0,design.n_TR,num=5)
+    scan_onsets = np.linspace(0,design.n_TR,num=n_run+1)
 
 
     # Test fitting with GP prior.
-    brsa = BRSA(GP_space=True,GP_inten=True,verbose=False,n_iter = 200)
+    brsa = BRSA(GP_space=True,GP_inten=True,verbose=False,n_iter = 200,rank=n_C)
 
     # test if the gradients are correct
-    XTY,XTDY,XTFY,YTY_diag, YTDY_diag, YTFY_diag, XTX, XTDX, XTFX = brsa._prepare_data(design.design_used,Y,n_T,n_V,scan_onsets)
+    XTY, XTDY, XTFY, YTY_diag, YTDY_diag, YTFY_diag, XTX, \
+        XTDX, XTFX, X0TX0, X0TDX0, X0TFX0, XTX0, XTDX0, XTFX0, \
+        X0TY, X0TDY, X0TFY, X0, n_run_returned, n_base =\
+            brsa._prepare_data(design.design_task,Y,n_T,scan_onsets)
+    assert n_run_returned == n_run, 'There is mistake in counting number of runs'
+    assert np.ndim(XTY) == 2 and np.ndim(XTDY) == 2 and np.ndim(XTFY) == 2,\
+        'Dimension of XTY etc. returned from _prepare_data is wrong'
+    assert np.ndim(YTY_diag) == 1 and np.ndim(YTDY_diag) == 1 and np.ndim(YTFY_diag) == 1,\
+        'Dimension of YTY_diag etc. returned from _prepare_data is wrong'
+    assert np.ndim(XTX) == 2 and np.ndim(XTDX) == 2 and np.ndim(XTFX) == 2,\
+        'Dimension of XTX etc. returned from _prepare_data is wrong'
+    assert np.ndim(X0TX0) == 2 and np.ndim(X0TDX0) == 2 and np.ndim(X0TFX0) == 2,\
+        'Dimension of X0TX0 etc. returned from _prepare_data is wrong'
+    assert np.ndim(XTX0) == 2 and np.ndim(XTDX0) == 2 and np.ndim(XTFX0) == 2,\
+        'Dimension of XTX0 etc. returned from _prepare_data is wrong'
+    assert np.ndim(X0TY) == 2 and np.ndim(X0TDY) == 2 and np.ndim(X0TFY) == 2,\
+        'Dimension of X0TY etc. returned from _prepare_data is wrong'
+    X0 = np.ones(n_T)
     l_idx = np.tril_indices(n_C)
+    # rank = n_C - 1
+    # idx_rank = np.where(l_idx[1] < rank)
+    # l_idx = (l_idx[0][idx_rank], l_idx[1][idx_rank])
```

Comment:
Please do not comment out unnecessary code; instead, remove it. There are other instances.